src/kudu/consensus/leader_election.cc

```diff
@@ -145,11 +145,116 @@ void FlexibleVoteCounter::FetchTopologyInfo(
   CHECK(le_quorum_requirement);
   voter_distribution->clear();
   le_quorum_requirement->clear();
-  // TODO(ritwikyadav): Populate the out params by reading config.
+
+  std::string leader_region = "";
+
+  // Step 1: Compute number of voters in each region.
+  // Also, computes the leader's region from its UUID.
+  std::set<std::string> unique_uuids;
+  for (const RaftPeerPB& peer : config_.peers()) {
+    // Only consider peers that are voters.
+    if (!peer.has_member_type() ||
+        peer.member_type() != RaftPeerPB::VOTER) {
+      continue;
+    }
+    CHECK(peer.has_permanent_uuid() && peer.has_attrs() &&
+        peer.attrs().has_region());
+    const std::string& region = peer.attrs().region();
+    if (unique_uuids.find(peer.permanent_uuid()) == unique_uuids.end()) {
+      int& count = LookupOrInsert(voter_distribution, region, 0);
+      count++;
+      unique_uuids.insert(peer.permanent_uuid());
+      if (peer.permanent_uuid() == leader_uuid_) {
+        leader_region = region;
+      }
+    }
+  }
+
+  CHECK(config_.has_commit_rule());
+
+  // A flag to denote if we could resolve a master.
+  // May involve cases of first election in the replicaset,
+  // last known master no longer part of the replicaset etc.
+  bool master_rule_unresolvable = false;
+
+  // Stores the requirement in the master region for a pessimistic
+  // computation of leader election quorum.
+  int master_region_req = -1;
+
+  // Step 2: Compute the leader election quorum from the corresponding
+  // data commit quorum. The requirement is for it to intersect with all the
+  // data commit quorums and all the leader election quorums.
+  for (const RegionRuleSpecPB& rule_spec :
+      config_.commit_rule().rule_conjunctions()) {
+    std::string region = rule_spec.region_identifier();
+    const std::string& commit_req = rule_spec.commit_requirement();
+    int commit_req_num = -1;
+    if (commit_req != "majority") {
+      // This will throw if the configuration is not properly defined.
+      commit_req_num = std::stoi(commit_req);
+    }
+
+    if (region == "master" && leader_region == "") {
+      master_rule_unresolvable = true;
+      master_region_req = commit_req_num;
+      continue;
+    } else {
+      region = leader_region;
+    }
+
+    // At this point region is resolved and we should be able to
+    // fetch its cardinalities.
+    // TODO(ritwikyadav): It is possible that all voters in a region
+    // have left the replicaset. Handle that case gracefully in a subsequent diff.
+    int total_voters = FindOrDie(*voter_distribution, region);
+    int region_majority = total_voters / 2 + 1;
+    if (commit_req == "majority") {
+      commit_req_num = region_majority;
+    }
+
+    // We chose the maximum of the two conditions required for intersection
+    // of leader election quorum with data commit quorum and for any two
+    // leader election quorums to intersect with one another.
+    int le_req_num = std::max(
```

Comment:
is this max operation really required?
if num voters in a region is 9, and some config says that it wants DQ = 7 votes,
then LE vote should be 9 + 1 - 7 = 3 ?